articles/zero_copy.md

```diff
@@ -0,0 +1,249 @@
+In this case, the user needs to call publish, then keep the message instance immutable until the middleware lets the user know that it is done with the loaned data.
+
+This is a narrow use case and will require additional interfaces to support, therefore it will be out of scope for this document, but it is mentioned for completion.
+
+### Subscription Use Cases
+
+There are two ways the users may take message instances from a subscription when data is available:
+
+1. Taking direction from the `Subscription` after polling it for data availability or waiting via a wait set.
+2. Using an `Executor`, which takes the data from the user and delivers it via a user-defined callback.
+
+Note: It is assumed that the user will be abe to take multiple messages at a time if they are available.
+
+In the first case, the user could choose to either:
+
+* Manage the memory for the message instance themselves, providing a reference to it, into which the middleware should fill the data.
+* Take one or more loaned messages from the middleware and return the loans later.
+
+In the second case, the user is delegating the memory management ot he client library via the `Executor`.The `Executor` may or may not borrow data from the middleweare, but the user callback does not care, so this can be considered an implementation detail.
+The user should be able to influence what the `Executor` does, and in the case that memory needs to be allocated, the user should be able to provide an allocator or memory management strategy which would influence the `Executor`'s behavior.
+
+## Requirements
+
+Based on the the use cases above, the general requirements are as follows:
+
+* Users must be able to avoid all memory operations and copies in at least one configuration.
+
+
+### Publisher Requirements
+
+* The user must be able to publish from messages allocated in their stack or heap.
+* The user must be able to get a loaned message, use it, and return it during publication.
+* When taking a loan, the middleware should not do anything "special", that is that the user must be able to influence the allocation.
+
+### Subscription Requirements
+
+The following requirements should hold whether the user is polling or using a wait set.
+
+* The user must be able to have the middleware fill data into messages allocated in the user's stack or heap.
+* The user must be able to get a loaned message from the middleware when calling take.
+* The user must be able to get a sequence of loaned messages from the middleware when calling take.
+* The loaned message or sequence must be returned by the user.
+* When taking a loan, the middleware should not do anything "special", that is that the user must be able to influence the allocation.
```

Comment:
Should that be when _not_ taking a loan? To my understanding the middleware should be completely in charge of allocations when taking a loan. However, if the middleware isn’t capable of zero copy, the user should have means to pass in custom allocators and thus control the allocation.

Reply:
I was a bit unclear on this as I was going through the previous notes.

Is the idea that users should _always_ take a loan, regardless of if the middleware supports it?  That would put the burden on the middleware for managing that memory.  Or should it be the case that there is no allocation when loans aren't supported?

Reply:
I think @Karsten1987 is correct, based on what I read in the use-case section above,

> the user is delegating the memory management of the client library via the Executor. The Executor may or may not borrow data from the middleware, but the user callback does not care, so this can be considered an implementation detail. The user should be able to influence what the Executor does, and **in the case that memory needs to be allocated**, the user should be able to provide an allocator or memory management strategy which would influence the Executor's behavior.

I've made bold the key phrase "in the case that memory needs to be allocated". My interpretation is the the client library will try to take a loan if possible, and if it is not possible then use any allocator/memory management strategy provided by the user.